tiledb/sm/cpp_api/vfs_experimental.h

```diff
@@ -145,14 +175,26 @@ class VFSExperimental {
       const Context& ctx,
       const VFS& vfs,
       const std::string& uri,
-      LsCallback cb) {
-    CallbackWrapperCPP wrapper(cb);
-    ctx.handle_error(tiledb_vfs_ls_recursive(
-        ctx.ptr().get(),
-        vfs.ptr().get(),
-        uri.c_str(),
-        ls_callback_wrapper,
-        &wrapper));
+      LsFileCallback file_cb,
+      std::optional<LsDirCallback> dir_cb = std::nullopt) {
+    if (dir_cb.has_value()) {
+      CallbackWrapperCPP wrapper(std::move(file_cb), dir_cb.value());
+      ctx.handle_error(tiledb_vfs_ls_recursive_v2(
```

Comment:
I could not think of a way around the `_v2` C API. For C++ the public API here selects the underlying _v2 or v1 API based on whether or not we have provided a directory filter.

If we don't want a v2, something that comes to mind is pulling out the directory filter and updating the documentation to state that the single callback should also handle directories.

A possible benefit to the v2 API is we could get recursive ListObjects performance with no prefixes collected by using the v1 API, while the v2 API would have some additional overhead for collecting the prefixes but it would be opt-in if prefixes are required. I have some booleans in place within the CallbackWrapper classes for this purpose, but they are not in use yet.